File: 백준/Silver/1874. 스택 수열/스택 수열.js
const readline = require("readline");

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

function solve(input) {
  const lines = input.trim().split("\n");
  const n = parseInt(lines[0]);

  const stack = [];
  let top = 0;
  let index = 1;
  let result = "";

  while (index <= n) {
    const value = parseInt(lines[index]);

    if (value > top) {
      for (let i = top; i < value; i++) {
        stack.push(i + 1);
        result += "+\n";
      }
      top = value;
    } else if (stack[stack.length - 1] !== value) {
      console.log("NO");
      return;
    }

    stack.pop();
    result += "-\n";
    index++;
  }

  console.log(result);
}

let input = "";

rl.on("line", (line) => {
  input += line + "\n";
});

rl.on("close", () => {
  solve(input);
});

File: 백준/Silver/1874. 스택 수열/스택 수열.py
import sys

def solve(lines):
    lines = lines.strip().split("\n")
    n = int(lines[0])

    stack = []
    top = 0
    index = 1
    result = ""

    while index <= n:
        value = int(lines[index])

        if value > top:
            for i in range(top, value):
                stack.append(i + 1)
                result += "+\n"
            top = value
        elif stack[-1] != value:
            print("NO")
            return

        stack.pop()
        result += "-\n"
        index += 1

    print(result)

input_lines = sys.stdin.readlines()
input_text = "".join(input_lines)
solve(input_text)
